src/main/java/io/lakefs/PathMapper.java

```diff
@@ -0,0 +1,311 @@
+
+    protected static final String MAPPING_CONFIG_PREFIX = "routerfs.mapping.";
+    protected static final String NON_DEFAULT_MAPPING_CONFIG_PATTERN = "^routerfs\\.mapping\\.([a-z0-9-_]*)\\.(\\d*)\\.(replace|with)";
+    protected static final String DEFAULT_MAPPING_CONFIG_PATTERN = "^routerfs\\.mapping\\.(.*default)\\.(replace|with)";
+    protected static final int NON_DEFAULT_TO_FS_IDX = 1;
+    protected static final int NON_DEFAULT_MAPPING_IDX_IDX = 2;
+    protected static final int NON_DEFAULT_MAPPING_TYPE_IDX = 3;
+    protected static final int DEFAULT_TO_FS_IDX = 1;
+    protected static final int DEFAULT_MAPPING_TYPE_IDX = 2;
+
+    enum MappingConfigType {
+        REPLACE,
+        WITH
+    }
+
+    private List<PathMapping> pathMappings;
+    private PathMapping defaultMapping;
+
+    public PathMapper(Configuration conf) throws IOException {
+        this.pathMappings = new ArrayList<>();
+        loadMappingConfig(conf);
+        if (defaultMapping == null) {
+            throw new IllegalArgumentException("Missing default mapping configuration, cannot initialize path mapper");
+        }
+    }
+
+    private void loadMappingConfig(Configuration conf) throws InvalidPropertiesFormatException {
+        List<MappingConfig> mappingConfiguration = parseMappingConfig(conf);
+        populatePathMappings(mappingConfiguration);
+
+        if (LOG.isDebugEnabled()) {
+            logLoadedMappings();
+        }
+    }
+
+    /**
+     * Pair mapping configuration that form a {@link PathMapping}, and populate the pathMapping list.
+     * Two mapping configuration are paired together in case they share the same toFsScheme and mappingIdx, but have
+     * different {@link MappingConfigType}.
+     * e.g. the following two mapping configuration for a path mapping:
+     * routerfs.mapping.lakefs.1.replace='^s3a://bucket/prefix'
+     * routerfs.mapping.lakefs.1.with='lakefs://example-repo/dev/prefix'
+     *
+     * @param mappingConfiguration the configurations to create path mapping from
+     */
+    private void populatePathMappings(List<MappingConfig> mappingConfiguration) {
+        List<MappingConfig> srcStyles = mappingConfiguration.stream()
+                .filter(mc -> mc.getType() == MappingConfigType.REPLACE).collect(Collectors.toList());
+
+        for (MappingConfig srcStyle : srcStyles) {
+            Optional<MappingConfig> matchingDstStyle = mappingConfiguration.stream()
+                    .filter(mc -> mc.getToScheme().equals(srcStyle.getToScheme()) && mc.getIndex() == srcStyle.getIndex()
+                            && mc.getType() == MappingConfigType.WITH).findFirst();
+            if (!matchingDstStyle.isPresent()) {
+                LOG.warn("Missing a mapping configuration, expected to find mapping named %s.%d.%s.",
+                        srcStyle.getToScheme(), srcStyle.getIndex(), MappingConfigType.WITH.name());
+                continue;
```

Comment:
What happens with this warning?  Shouldn't we throw an exception and refuse to run?  (This is Spark: nobody will see a warning.)